### PR TITLE
Remote otel collector configuration

### LINKED ISSUE #933: Support send data to remote Otel collector
This issue is keep track of a subset of the changes necessary to allow Kubewarden send data to remote  OpenTelemetry (Otel) collectors. As described in #931.

Currently, the controller is able to send data to a Otel collector running as sidecar.  It's necessary to change exporter in the Kubewarden controller to send data to a remote Otel collector and properly configure the policy servers to do the same

### Acceptance criteria
- [ ] Add a flag to signal when the Otel collector is running as sidecar
- [ ] Change the [code](https://github.com/kubewarden/kubewarden-controller/blob/main/internal/metrics/metrics.go) used to configure the Otel collector to consider the information from the environment variable.
- [ ] Replicate the Otel configuration defined in the controller into the policy servers deployment. Including:
  - [ ] Otel collector endpoint
  - [ ] Otel TLS configuration(e.g. insecure connections and/or certificate to be used)

### PR DESCRIPTION
**Pull request by jvanz**
## Description

Updates the collector to allow the communication with a remote OpenTelemetry collector. Also updates the policy server reconciler to replicate the same configuration in the policy server deployment enabling it to send data to the same remote Otel collector.

Fix #933 

## Tests

> [!WARNING]  
> This PR was tested together with the changes from [this](https://github.com/kubewarden/helm-charts/pull/581) Helm chart changes

Create a simple cluster with required dependencies:

```
k3d cluster create 

helm repo add jetstack https://charts.jetstack.io --force-update
helm upgrade -i --wait cert-manager jetstack/cert-manager \
        -n cert-manager --create-namespace \
        --set crds.enabled=true

helm repo add --force-update open-telemetry https://open-telemetry.github.io/opentelemetry-helm-charts
helm upgrade -i --wait my-opentelemetry-operator open-telemetry/opentelemetry-operator \
        --set "manager.collectorImage.repository=otel/opentelemetry-collector-contrib" \
        -n open-telemetry --create-namespace
```

I've tested this changes with this Otel collector configuration

```yaml
# cert-manager resources
apiVersion: cert-manager.io/v1
kind: Certificate
metadata:
  name: my-client-certificate
  namespace: kubewarden
spec:
  dnsNames:
  - my-collector-collector.kubewarden.svc
  - my-collector-collector.kubewarden.svc.cluster.local
  issuerRef:
    kind: Issuer
    name: my-client-selfsigned-issuer
  secretName: my-client-cert
---
apiVersion: cert-manager.io/v1
kind: Issuer
metadata:
  name: my-client-selfsigned-issuer
  namespace: kubewarden
spec:
  selfSigned: {}
---
apiVersion: cert-manager.io/v1
kind: Certificate
metadata:
  name: my-certificate
  namespace: kubewarden
spec:
  dnsNames:
  - my-collector-collector.kubewarden.svc
  - my-collector-collector.kubewarden.svc.cluster.local
  issuerRef:
    kind: Issuer
    name: my-selfsigned-issuer
  secretName: my-server-cert
---
apiVersion: cert-manager.io/v1
kind: Issuer
metadata:
  name: my-selfsigned-issuer
  namespace: kubewarden
spec:
  selfSigned: {}
---
apiVersion: opentelemetry.io/v1beta1
kind: OpenTelemetryCollector
metadata:
  name: my-collector
  namespace: kubewarden
spec:
  mode: deployment # This configuration is omittable.
  volumes:
    - name: server-certificate
      secret:
        secretName: my-server-cert
    - name: client-certificate
      secret:
        secretName: my-client-cert
  volumeMounts:
    - name: server-certificate
      mountPath: /tmp/etc/ssl/certs/my-server-cert
      readOnly: true
    - name: client-certificate
      mountPath: /tmp/etc/ssl/certs/my-client-cert
      readOnly: true
  config:
    receivers:
      otlp:
        protocols:
          grpc:
            tls:
              cert_file: /tmp/etc/ssl/certs/my-server-cert/tls.crt
              key_file: /tmp/etc/ssl/certs/my-server-cert/tls.key
              client_ca_file: /tmp/etc/ssl/certs/my-client-cert/ca.crt
              # ca_file: /tmp/etc/ssl/certs/my-server-cert/ca.crt
              # client_ca_file_reload: true
    processors: {}
    exporters:
      debug:
        verbosity: normal
      prometheus:
        endpoint: ":8080"
    service:
      pipelines:
        metrics:
          receivers: [otlp]
          processors: []
          exporters: [debug, prometheus]
        traces:
          receivers: [otlp]
          processors: []
          exporters: [debug]

```

### CODE REVIEW

File: cmd/main.go
Comment (jvanz) [inline cmd/main.go:103]: I would like to try to simplify all these CLI flags.

File: internal/controller/policyserver_controller_deployment.go
Comment (flavio) [inline internal/controller/policyserver_controller_deployment.go:312]: Can't we add the code on the policy server side to handle this as an env variable too? Just for the sake of consistency and to keep this code simpler (not have to maintain the cognitive overload of what is to be passed as flag vs cli flag).

CC @fabriziosestito

File: internal/controller/policyserver_controller_deployment.go
Comment (flavio) [inline internal/controller/policyserver_controller_deployment.go:327]: This is fragile. Let's coordinate with @fabriziosestito to handle it better. Having the `http` hard coded is also bad (no worries, I understand this is a draft); we have to take into account the `https` scenario

File: cmd/main.go
Comment (fabriziosestito) [reply cmd/main.go:103]: On second thought, I think we can remove those if we are using the `OTEL_*` env variables.

File: internal/controller/policyserver_controller_deployment.go
Comment (fabriziosestito) [reply internal/controller/policyserver_controller_deployment.go:312]: > Can't we add the code on the policy server side to handle this as an env variable too

I think this is the way to go. I misunderstood, as I thought we were exposing only the endpoint and the certificate settings. 

To summarize, we should only use environment variables.
 I am updating the policy server to mimic the missing upstream logic for the certificates env vars.

File: internal/controller/policyserver_controller_deployment.go
Comment (flavio) [inline internal/controller/policyserver_controller_deployment.go:249]:
```suggestion
	// If the otel sidecar is disabled, we  need to configure the policy
```

File: internal/controller/policyserver_controller_deployment.go
Comment (flavio) [inline internal/controller/policyserver_controller_deployment.go:253]:
```suggestion
	// necessary to change the `PolicyServer` CRD.
```

**Review by flavio — CHANGES_REQUESTED** [review]
It looks good. I left some minor comments, some of them are not mandatory.

Please address the linter warnings before merging the PR

File: cmd/main.go
Comment (fabriziosestito) [reply cmd/main.go:149]: +1

File: internal/controller/policyserver_controller_deployment.go
Comment (fabriziosestito) [reply internal/controller/policyserver_controller_deployment.go:267]: - linter is complaining about nested ifs, maybe we could create a separate method.

- `r.OtelCertificateSecret != ""` is more idiomatic

File: internal/controller/policyserver_controller_deployment.go
Comment (fabriziosestito) [reply internal/controller/policyserver_controller_deployment.go:283]: same as above, empty strings should be checked by `!= ""`

File: internal/controller/policyserver_controller_deployment.go
Comment (jvanz) [reply internal/controller/policyserver_controller_deployment.go]: Yes, I was thinking about that as well. Another option is to add some validation in the webhook to prevent the users from deploying a policy server with a configuration that can cause issues. Therefore, we can at least tell them why the controller cannot deploy the policy server. Instead of allow they to deploy and notice that something is not working as they expected afterwards.

File: internal/controller/policyserver_controller_deployment.go
Comment (viccuad) [inline internal/controller/policyserver_controller_deployment.go:354]: does this need to be idempotent? (same for `otelClientCertificateSecret`)

File: Tiltfile
Comment (viccuad) [reply Tiltfile]: (unresolving) I have the same doubt here, can we continue and skip if this weren't provided instead of fail?

Comment (jvanz) [reply Tiltfile]: The last version of this PR already continue if this configuration is not defined.

Comment (viccuad) [reply Tiltfile]: You are right, I see now that `additionalHelmInstallationArgs` defaults to `{}` and one can comment `controller_values_file` from the example.

Resolving.

File: internal/controller/policyserver_controller_deployment.go
Comment (flavio) [reply internal/controller/policyserver_controller_deployment.go:354]: This is idempotent, give we always start from an empty Deployment and add all the data defined inside of the PolicyServer `spec`, plus all the values we deem required.

@kubewarden/kubewarden-developers am I right?

**Review by flavio — APPROVED** [review]
I think this looks good. Please wait for everybody's approvals before merging

File: Tiltfile
Comment (fabriziosestito) [inline Tiltfile:42]: nit: let's use snake case for variable names https://peps.python.org/pep-0008/#function-and-variable-names

**Review by fabriziosestito — APPROVED** [review]

**Review by viccuad — APPROVED** [review]
LGTM!

**Comment by flavio** [comment]
Merging, addressing the code coverage is tough, we will track this technical debt and address that after the 1.20 release

**Pull request merged**
